### ISSUE TITLE
Share driver between multiple instances of SearchProvider

### ISSUE THREAD

**Issue by FokkeZB** [elastic/search-ui#409]
The children of `SearchProvider` will only render when the driver has been initialized, which on SSR, only happens client-side:

https://github.com/elastic/search-ui/blob/81bc09c90e3ca73e5565c505084ec4613f463099/packages/react-search-ui/src/SearchProvider.js#L52-L54

We use `SearchBox` on every page and on the search results page we of course also use `SearchResults` (and friends). On the search page we wrap pretty much everything in `SearchProvider`, but on the other pages we can't because that would hurt our SEO (the server-side rendered page would be close to empty).

So what we did is wrap just `SearchBox`, but as this lives under a component also used on the search page you end up with two nested `SearchProvider` and we noticed they both make the same requests when on the search page.

To work around this for now, we're using this instead of the `SearchProvider` directly:

```jsx
const SearchProviderOnce = (props: Props) => {
  const { children } = props;

  return (
    <SearchContext.Consumer>
      {value => value ? children : (
        <SearchProvider config={config}>{children}</SearchProvider>
      )}
    </SearchContext.Consumer>
  );
};
```

This ensures that we only wrap once.

A more clean solution would be great, e.g. by sharing all `SearchProvider` instances (at least those with the same `config` property) or their drivers.

**Comment by FokkeZB** [comment]
Yes, we'd still very much like to get rid of our `SeasrchProviderOnce`.

**Comment by JasonStoltz** [comment]
Thanks for the response @FokkeZB!

**Comment by JasonStoltz** [comment]
Just re-reading this. This is interesting. You should be able to have 2 `SearchProviders` on a page and they should act in complete isolation. The only overlap I believe they would have is via `trackUrlState`.

So if you have two SearchProviders on a page and they both are not using `trackUrlState`, they should be fine.

However, it sounds like you might be saying that you have one nested inside the other. Is that the case... that you have one SearchBox nested under the other?

Thanks

**Comment by FokkeZB** [comment]
They're nested indeed. On all pages under https://zapier.com/help we have a search input with autocomplete in our header that is part of a layout component. So we wrap the layout in the search provider.

However, on the results page (e.g. https://zapier.com/help/search?q=pizza) we pass children to the layout component that also need access to the provider. So we wrap it in the provider again.

We _could_ make the `children` of the layout component a function that we call with the context that we need, but then we'd need to pass this down to all components that the children use as well.

If you have other ideas on how to construct this, we'd love to hear. The complexity that SSR adds certainly doesn't help 😉

**Comment by JasonStoltz** [comment]
Hey @FokkeZB, I believe I may have resolved a bit unintentionally, and it never occurred to me to follow up here.

https://github.com/elastic/search-ui/commit/2bf0a330ebbc755ac16d5a9643a4d88d3d38265d

I made it so that a `SearchDriver` can be instantiated independently of a `SearchProvider`, and then passed in as a prop to `SearchProvider`.

This essentially means that you can have multiple `SearchProvider`s on a page that all work as a single unit.

Here is a working example of this in action:
https://codesandbox.io/s/search-ui-national-parks-example-forked-4e5ti?file=/src/App.js

```
const driver = new SearchDriver(config);

export default function App() {
  return (
    <div className="App">
      <Header>
          <SearchProvider driver={driver}>
            <SearchBox />
          </SearchProvider>
      </Header>
      <Body>
          <SearchProvider driver={driver}>
            <Results />
          </SearchProvider>       
      </Body>
    <div />
   )
```

**Comment by FokkeZB** [comment]
@JasonStoltz that looks like something we could use indeed. We'd wrap only the body of the search results page in the provider and keep doing the same for the header with search field found on every page - both using the same driver reference.

Thanks!
